### DISCUSSION TITLE
locally loading exported .zip project?

### DISCUSSION THREAD

**Discussion by michaelkowalczyk** [LiaScript/LiaScript#187] [General]
Each time when I like to show a project preview to a client or collaborator I need to do following steps. 
- export the project as .zip (.md file and uploaded media)
- upload and unzip everything on a webserver (via ftp in my case)* 
- send the full course loading url or ask client to enter the .md url on https://liascript.github.io/course/

Would it be possible to make the process quicker and have a way to load the project .zip locally in the browser?

Taking all files online is just needed when publishing the course at the final stage.

*working with free version of github as place to serve the project files is only possible when they are made public. When private the .md files url is not accessible. 

The collaborative? WebRTC or Websocket functions are currently a mystery for me. Nothings comes up when searching for the terms in the documentation course https://liascript.github.io/course/?https://raw.githubusercontent.com/liaScript/docs/master/README.md#1 

When using "Export to Github Gist" the externally embedded media files like "youtube" and "h5p" element do not appear. 

Best regards
Michael

**Comment by andre-dietrich** [comment]
An easy way to directly share your content, for demonstration purposes, would be to use WebTorrent. I wrote a little post on this:

https://liascript.github.io/blog/sharing-content-via-webtorrent/

This would only require to download and extract the zip and then to move all files to instant.io ...

---

If you have ownCloud or Nextcloud, then you should also be able to upload your content there and simply link it, as with GitHub. With Dropbox, this is unfortunately not possible anymore ...

---

If you have only an example, you could try to share your content directly from the editor "Share via data-URI" this will encode the entire README into the URL, which can then be shared with your colleagues. The longer the content, the longer the URL. This works also with the editor, by using snap-shot-URL, this will encode the content and open the editor ...

---

However, and as you have mentioned, you can go online with your editor. You can try this out with different browsers. If you go online with WebRTC, then the URL changes. Copy this URL into the other Browser (don't use the same browser twice), the states will be synced. Hence, if you have a couple of multimedia-files uploaded, the synchronization might take some time ... In this case you can now work collaboratively on the same document, all changes will be transmitted from one browser to the other ... WebSocket, simply uses another server as a relay, but does not store any content. All synchronization happens securely only between the browsers as endpoint ...

---

Uploading multimedia to gist is unfortunately not possible by using the API. Although, I can upload multiple files, only text files are accepted. I also tried base64 encoding / data-protocol, of images which means, the images are stored as hell long URLs within the text. But, this is also prohibited and blocked by GitHub.

I hope that we can later connect such projects from the live-editor and directly connect them with GitHub...

**Reply by michaelkowalczyk** [reply]
> An easy way to directly share your content, for demonstration purposes, would be to use WebTorrent. I wrote a little post on this:
> https://liascript.github.io/blog/sharing-content-via-webtorrent/
> This would only require to download and extract the zip and then to move all files to instant.io ...
> 

Sounds like a novel idea which I tried out. My browser at the uni network encountered following, probably security related blockades "Error connecting to wss://tracker.btorrent.xyz " and "Error connecting to wss://tracker.openwebtorrent.com"

> If you have ownCloud or Nextcloud, then you should also be able to upload your content there and simply link it, as with GitHub. With Dropbox, this is unfortunately not possible anymore ...

The Uni uses Nextcloud.
This works only partially AND one has to manually modify the shared url of .md file by adding "/download/readme.md" at the end of the shared link. Then the readme is loaded when pasting the link on https://liascript.github.io/ 
What still does not work, is the display of the images which were uploaded to the project, because the link is for the readme.md only.
https://cloud.uni-konstanz.de/index.php/s/nQ7nGfcWHqyre9z/download/README.md 
Here is the link to the image file https://cloud.uni-konstanz.de/index.php/s/ZRi38HZY43FWwgd/download/257eb7c12fc717cb647bdd19b7207cb5c5df1277.png

![Screenshot - 20_08_2024 , 10_37_41 002](https://github.com/user-attachments/assets/1d606d97-74d1-401c-989e-b3b48bcf8f80)

> 
> If you have only an example, you could try to share your content directly from the editor "Share via data-URI" this will encode the entire README into the URL, which can then be shared with your colleagues. The longer the content, the longer the URL. This works also with the editor, by using snap-shot-URL, this will encode the content and open the editor ...

I used the same example and shared the longes data URI

URL-length: 434 bytes
https://liascript.github.io/course/?data:text/plain,%3C!--%0D%0Aversion%3A%20%200.0.1%0D%0A%0D%0Acomment%3A%20%20Test%20Nextcloud%20Sharing%20for%20LiaScript%0D%0A%20%20%20%20%20%20%20%20%20%20%0D%0Adate%3A%20%20%20%20%2020%2F08%2F2024%0D%0A%0D%0A--%3E%0D%0A%0D%0A%23%20Test%20Nextcloud%20Sharing%20for%20LiaScript%0D%0A%0D%0A%0D%0AImage%20uploaded%20in%20live%20editor%0D%0A%0D%0A!%5B%5D(257eb7c12fc717cb647bdd19b7207cb5c5df1277.png) 

When I click on it the course loads in the browser but the images is still missing. 
Should this actually work with images too?

![Screenshot - 20_08_2024 , 10_48_29](https://github.com/user-attachments/assets/ffe62ec3-1386-4bac-868e-10dfd254581b)

> 
> However, and as you have mentioned, you can go online with your editor. You can try this out with different browsers. If you go online with WebRTC, then the URL changes. Copy this URL into the other Browser (don't use the same browser twice), the states will be synced. Hence, if you have a couple of multimedia-files uploaded, the synchronization might take some time ... In this case you can now work collaboratively on the same document, all changes will be transmitted from one browser to the other ... WebSocket, simply uses another server as a relay, but does not store any content. All synchronization happens securely only between the browsers as endpoint ...

That is a great feature. As I understand this works only when both clients are online?
Due to time constraints this is not always possible and the client would not need to see the live editor either.

> 
> Uploading multimedia to gist is unfortunately not possible by using the API. Although, I can upload multiple files, only text files are accepted. I also tried base64 encoding / data-protocol, of images which means, the images are stored as hell long URLs within the text. But, this is also prohibited and blocked by GitHub.
> 
> I hope that we can later connect such projects from the live-editor and directly connect them with GitHub...

Thanks for all the hard work you already put into the project!
Best regards, Michael

**Comment by andre-dietrich** [comment]
Concerning WebTorrent, some of these "trackers" might fail, I think if the network administrator is not too restrictive, then you will get a connection.

----

The thing with the NextCloud is bad news ... LiaScript uses the URL of the README as the base URL and sets the address of images, etc. relative to this. If the path of the README is somehow encoded within the query-parameters, then this approach fails. Especially if keys are used ... Some time ago, Dropbox offered such simple download-links, but they removed it ...

Thanks for trying out... I will have to check, if this can be solved somehow ...

---

Embedding images into the URL does not work at the moment. I think this will also blow-up the URL length ... After a few thousand characters, GitHub will block it ... 

---

Syncing should work in theory also if you are offline and later come back online ... so two authors should be able to synchronize their work also after offline times ... We use Yjs for this task. A library for conflict-free-replicated-data types. Also in theory they should always merge, thus, all peers will end up with the same version, but Yjs can also be very strict, when merging ... If you predefine your areas:

``` markdown
# course

--------------

author A

# Sec 1

....

--------------

author B

# Sec n
```

Then there should not be any problems when merging your states. Only if changes are too chaotic, it might not look as expected ...

You can try this out with different browser ...

----

Due to this conversation, I experimented a bit with file-uploads. Actually, it should be not too complicated to implement something like this, which would allow uploading folders or zip-files... I think, this would be an interesting feature also for offline-learning ...

**Reply by michaelkowalczyk** [reply]
Thanks for the reply and explanations Andre.
A zip & share function could fit very well into liascripts offline approach.
Additinally a split in "share, & collaborate" and "share & show" could be helpful to meet user specific needs.

**Comment by andre-dietrich** [comment]
@michaelkowalczyk ... If you load the liascript index https://liascript.github.io/course , you will have now the possibility to upload projects directly from your device or from a zip file ... I hope the descriptions is self-explanatory ;-)

**Reply by michaelkowalczyk** [reply]
Wow, thank you for implementing a local course zip usage option Andre. This definetly enhances Liascrip ability for oer course sharing!

I had a look at the /course page layout from a usability point of view and also from the descriptive text. I documented the findings (or rather a open wishlist) on the following image:

![Screenshot - 30_09_2024 , 10_17_48](https://github.com/user-attachments/assets/248b918a-000c-4045-9eb2-3610765f40cb)

Best regards.
Michael

**Reply by michaelkowalczyk** [reply]
Also did you think about making a functional connection between the the "Open-course" page and the "LiveEditor"? 

Right now the situation is disconnected
- LiveEditor (edit) -> export project > open on course page (either load via url or import project zip)
- course page (view only) --> there is no way to open in LiveEditor (although all files are technically in the local browser cache?)

Problem
- Once a project (including embeded media files) is deleted from the local LiveEditor page one can not easily restore the project and edit it again. It is only possible if the used media files are embedded from fixed urls. Then one still needs to manually copy and paste the README.md text into the LiveEditor. 

I could imagine that at the end there is only need for one course page where courses are either opened to view or edit.

Best regards
Michael

**Reply by andre-dietrich** [reply]
Directly connecting the course and editor will be technically hard, these are separated projects with which use their own internal structure to store/cache the content ... The editor uses Yjs that allows for cooperative editing ... The editor was originally an attempt to let users play with examples, we thought that they would use VSCode or Atom or another editor. Originally, we were not aware that entire files and videos can be stored within the browsers internal database. We will improve this editor further to reduce the initial learning curve... But the idea sounds interesting to connect both applications, I will have to check if this is even possible atm....

But uploading a zip into the editor is definitely a good idea. This can be tackled when the tree-view for folders is implemented. At the moment, every file is stored only with its hash-value as its name. The problem is, that custom zip-files do have their own internal folder-structure ... 

To clarify some parts of the index:

7. The reset-button resets the entire internal state, which includes quizzes, surveys, tasks and codes. These will then be deleted
8. The different versions prevent old states from being deleted, if a new major-version of the document is loaded, which might have been restructured totally. So the user can always go back to the previous version, and go on with learning if he or she likes the old version better (I hope this makes sense ;-) ) ... And the different versions are used, when the course is loaded via a URL, this URL is used as the main reference. In a zip-file we thought, that this might be difficult, since we identify a course based on the file name for example ... 

1. & 2. What would you suggest as a more specific text? I tried to be as general as possible, so that you can upload a single zip-file, select multiple files (readme, images, videos, etc.) or an entire folder.

5., 4. and 3. are good points ... I will take them into account 

Filtering is also a good point ... originally I wanted to use a text-search, but since it is Markdown, I did not know how to show the text-snippet-results properly ... this would be very interesting ... and alternative that we had in mind, was something like a kanban-board, where the user can order and group the courses according to their own needs ... Probably I should resurrect this feat/branch ... We were also experimenting with graph-representations of content or how can documents be combined:

https://www.youtube.com/watch?v=ssb7e4Oh9As

Thanks for the suggestions ;-)

**Reply by michaelkowalczyk** [reply]
Hallo Andre, glad I can help with suggestions for LS.
>
> Directly connecting the course and editor will be technically hard, these are separated projects with which use their own internal structure to store/cache the content ... The editor uses Yjs that allows for cooperative editing ... The editor was originally an attempt to let users play with examples, we thought that they would use VSCode or Atom or another editor. Originally, we were not aware that entire files and videos can be stored within the browsers internal database. We will improve this editor further to reduce the initial learning curve... But the idea sounds interesting to connect both applications, I will have to check if this is even possible atm....
>

Why even bother using an external Editor when the LiveEditor sits in the browser and renders the result already.

For easy reference I give the LS parts descriptive names. 
- LSE - LiaScript Editor (browser based offline editor with preview + LSC course export, etc...)
- LSC - LiaScript Course (a markup formated text file (.md) + eventually embedded media (then .zip) created and exported within LSE. (also by other markup editors)
- LSV - LiaScript Viewer (the view which load from /courses)

- LSE:  I ask myself if there is even the necessity for having a seperate LSV /courses page at all. The LSE has a preview button AND a basic course overview page to display and load courses.  If it would be possible to implement some kind of author lock trigger (password?) to enable or disable editing, one could tell LSE to display the result only or enable the editor. 

> 
> But uploading a zip into the editor is definitely a good idea. This can be tackled when the tree-view for folders is implemented. At the moment, every file is stored only with its hash-value as its name. The problem is, that custom zip-files do have their own internal folder-structure ...
> 

yes, loading up projects zip instead of pasting the text and reuploading the local media into the editor could help.

> To clarify some parts of the index:
> 
> 7. The reset-button resets the entire internal state, which includes quizzes, surveys, tasks and codes. These will then be deleted.

Now its clear. 

> 8. The different versions prevent old states from being deleted, if a new major-version of the document is loaded, which might have been restructured totally. So the user can always go back to the previous version, and go on with learning if he or she likes the old version better (I hope this makes sense ;-) ) ...And the different versions are used, when the course is loaded via a URL, this URL is used as the main reference. In a zip-file we thought, that this might be difficult, since we identify a course based on the file name for example ...

Ok so a zip. project has only one version /  url project can have two or more?

> 9. & 2. What would you suggest as a more specific text? I tried to be as general as possible, so that you can upload a single zip-file, select multiple files (readme, images, videos, etc.) or an entire folder.

Overall the introduction of the page could benefit from a clearer description (esp. for new website viewers) what is meant and happening there. 

Heres my try:

On this page you can select a LiaScript Course (a markdown formatted text .md file) and open it in the LiaScript Viewer.

To open a LiaScript Course either select the url of the course or upload the course zip or the directory with individual files. 

* Load from URL (pointing to the the .md file) [...} 
   * (... search for free LiaScript courses and related material on [GitHub](https://github.com/topics/liascript _<- here you could add that one has to manually find and download the readme.md file_ ) 
* Import File(s) / Zip 
* Import Directory

To build your own LiaScript Course try our LiaScript Editor (add link).


> 
> 5., 4. and 3. are good points ... I will take them into account
> 
> Filtering is also a good point ... originally I wanted to use a text-search, but since it is Markdown, I did not know how to show the text-snippet-results properly ... this would be very interesting ... and alternative that we had in mind, was something like a kanban-board, where the user can order and group the courses according to their own needs ... Probably I should resurrect this feat/branch ... We were also experimenting with graph-representations of content or how can documents be combined:
> 
> https://www.youtube.com/watch?v=ssb7e4Oh9As
> 

One thing I noticed that the most recent open course is then displayed on the top of the list. This can become a bit tricky to follow up when different courses and versions pile up in the view. The sorting could help keept them ordered,

The same happens in the LSE /courses overview page and I succeeded to accidently delete my most recent LiveEditor course from the browser.

I think it would be overall better when the order of courses remains static or can be sorted manually. 

> Thanks for the suggestions ;-)

You are welcome. I try to improve the understanding of LS to new users and improve web usability :)

**Comment by andre-dietrich** [comment]
@michaelkowalczyk ... just updated the zip-loading functionality, which now allows uploading a zip-file to dropbox, github, or onedrive ... the Interpreter will now detect if it is a zip-file and apply the same methods for extracting and storing the content ... still a bit experimental and url-translation is done automatically ;-)
